NBitcoin.Altcoins/Stratis.cs

```diff
@@ -0,0 +1,660 @@
+					VarInt txCount = new VarInt(0);
+					stream.ReadWrite(ref txCount);
+
+					// Stratis specific addition - unknown usage see Stratis source.
+					stream.ReadWrite(ref txCount);
+				}
+
+				#endregion
+			}
+
+			public override void ReadWriteCore(BitcoinStream stream)
+			{
+				if (stream.Serializing)
+				{
+					var heardersOff = Headers.Select(h => new BlockHeaderWithTxCount(h)).ToList();
+					stream.ReadWrite(ref heardersOff);
+				}
+				else
+				{
+					Headers.Clear();
+					List<BlockHeaderWithTxCount> headersOff = new List<BlockHeaderWithTxCount>();
+					stream.ReadWrite(ref headersOff);
+					Headers.AddRange(headersOff.Select(h => h._Header));
+				}
+			}
+		}
+
+		public class StratisBlockSignature : IBitcoinSerializable
+		{
+			protected bool Equals(StratisBlockSignature other)
+			{
+				return Equals(signature, other.signature);
+			}
+
+			public override bool Equals(object obj)
+			{
+				if (ReferenceEquals(null, obj)) return false;
+				if (ReferenceEquals(this, obj)) return true;
+				if (obj.GetType() != this.GetType()) return false;
+				return Equals((StratisBlockSignature)obj);
+			}
+
+			public override int GetHashCode()
+			{
+				return (signature?.GetHashCode() ?? 0);
+			}
+
+			public StratisBlockSignature()
+			{
+				this.signature = new byte[0];
+			}
+
+			private byte[] signature;
+
+			public byte[] Signature
+			{
+				get
+				{
+					return signature;
+				}
+				set
+				{
+					signature = value;
+				}
+			}
+
+			internal void SetNull()
+			{
+				signature = new byte[0];
+			}
+
+			public bool IsEmpty()
+			{
+				return !this.signature.Any();
+			}
+
+			public static bool operator ==(StratisBlockSignature a, StratisBlockSignature b)
+			{
+				if (System.Object.ReferenceEquals(a, b))
+					return true;
+
+				if (((object)a == null) || ((object)b == null))
+					return false;
+
+				return a.signature.SequenceEqual(b.signature);
+			}
+
+			public static bool operator !=(StratisBlockSignature a, StratisBlockSignature b)
+			{
+				return !(a == b);
+			}
+
+			#region IBitcoinSerializable Members
+
+			public void ReadWrite(BitcoinStream stream)
+			{
+				stream.ReadWriteAsVarString(ref signature);
+			}
+
+			#endregion
+
+			public override string ToString()
+			{
+				return Encoders.Hex.EncodeData(this.signature);
+			}
+		}
+
+		/// <summary>
+		/// A POS block header, this will create a work hash based on the X13 hash algos.
+		/// </summary>
+		public class StratisBlockHeader : BlockHeader
+		{
+			const int CurrentVersion = 7;
+
+			private readonly X13 x13;
+
+			public StratisBlockHeader() : base()
+			{
+				this.x13 = new X13();
+			}
+
+			protected internal override void SetNull()
+			{
+				nVersion = CurrentVersion;
+				hashPrevBlock = 0;
+				hashMerkleRoot = 0;
+				nTime = 0;
+				nBits = 0;
+				nNonce = 0;
+			}
+
+			private byte[] CalculateHash(byte[] data, int offset, int count)
+			{
+				byte[] hashData = data.Skip(offset).Take(count).ToArray();
```

Comment:
I would change this and use the extension method `SafeSubarray` here. In this case it would look like `byte[] hashData = data.SafeSubarray(offset, count);`